openlibrary/macros/AffiliateLinks.html

```diff
@@ -1,48 +1,60 @@
 $def with (page, opts)
 
-$ prices = opts.get('prices')
-$ isbn = opts.get('isbn', '')
-$ asin = opts.get('asin', '')
-
-$ bwb_affiliate_url = 'https://www.anrdoezrs.net/links/%s/type/dlg/http://www.betterworldbooks.com/' % affiliate_id('betterworldbooks')
-$ amazon_affiliate_url = 'https://www.amazon.com/dp/%s/?tag=%s'
-
-$ bwb_price = None
-$ amz_price = None
+$code:
+  prices = opts.get('prices')
+  isbn = opts.get('isbn', '')
+  asin = opts.get('asin', '')
+
+  bwb = {
+    'key': 'betterworldbooks',
+    'analytics_key': 'BetterWorldBooks',
+    'name': _('Better World Books'),
+    'link': 'https://www.anrdoezrs.net/links/%s/type/dlg/http://www.betterworldbooks.com/%s' % (
+      affiliate_id('betterworldbooks'),
+      ('-id-%s.aspx' % isbn) if isbn else ('search/results?q=' + page.title)
+    ),
+    'price_note': _(' - includes shipping')
+  }
+
+  amazon = {
+    'key': 'amazon',
+    'analytics_key': 'Amazon',
+    'name': _('Amazon'),
+    'link': 'https://www.amazon.com/dp/%s/?tag=%s' % (asin or isbn, affiliate_id('amazon')),
+  } if (asin or isbn) else None
+
+  bookshop = {
+    'key': 'bookshop-org',
+    'analytics_key': 'BookshopOrg',
+    'name': _('Bookshop.org'),
+    'link': 'https://bookshop.org/a/%s/%s' % (affiliate_id('bookshop-org'), isbn),
+  } if isbn else None
+
```

Comment:
Ideally, this would all get moved inside of core/vendors.py; leaving as a suggestion, not blocker.